[JIRA-2796] Uploader now adds "item" field to JSON output, identifying the purpose of the upload.

RonConescu opened the pull request:
The point of this branch:  changed some +upload methods to require a "taskIdentifier" string, as requested by the server team, which shows up in the "item" field in the "info.json" file.  Also requires a "taskRunUuid," because elsewhere in the app, those two fields are tied together -- one says what we're uploading (like "monthly heart-health survey"), the other says which copy of that data is being uploaded.

Stuff I fixed as a side effect:
- enhanced the matching documentation
- fixed the MedicationTracker's Prescription class to use the new +upload method correctly
- changed literals to constants in the MedicationTracker's Prescription class

Testing procedures:

This was an incremental change to an existing "uploader" component.  The change involved adding two key-value pairs to a JSON manifest file in the upload. To test it, I wrote three methods that generated plain-text files and uploaded them, passing various combinations of following values for those two new fields:
- normal values, like "test_file.csv" and "happyFileName"
- nil
- UIUDs, both as NSUUID objects and as strings
- strings with "garbage" data, including randomly-typed characters, lots of spaces, backslash-n characters, tabs, and embedded quotation marks

...and verified that those strings ended up in the JSON file as intended -- either as the specified string or a JSON "null" value.

In addition, this change affected a component which uses the uploader: the "medication tracker." So after the change, I tested the medication tracker in the 3 situations which generate an upload:
- creating a prescription
- taking a pill for a prescription
- canceling the prescription

I ran those tests twice: once using the code from before the change, once using the code from after the change. Then I compared the "before" output for each situation to the "after" output for that same situation. Except for the two key-value pairs added as a result of this change, the output is identical, as it should be.

Commit:
[JIRA-2796] Uploader now adds "item" field to JSON output, identifying the purpose of the upload.

This commit changes the new +upload methods to require a "taskIdentifier" string, as requested by the server team.  Also requires a "taskRunUuid," because elsewhere in the app, those two field seem be tied together -- one says what we're uploading (like "monthly heart-health survey"), the other says which copy of that data is being uploaded.

Stuff I fixed as a side effect:
- enhanced matching documentation
- fixed the MedicationTracker's Prescription class to use the new +upload method correctly (or else the app wouldn't compile)

Commit:
[JIRA-2796] Declaring public static constants correctly.

Updating these constant declarations to use the correct, safer technique.

Commit:
[JIRA-2796] Made the ArchiverAndUploader publicly visible.

...so that the top-level apps can see it.  Verified with a local build of Cardio.

jwe-apple commented:
Could you also address the testing carried out in your PR or in the commit messages?

jwe-apple commented:
@RonConescu Are you watching this repo, or do you need specific notification?

Commit:
[JIRA-2796] Addressing comments from pull request.

- unified init methods
- clarified documentation
- changed literals to constants

jwe-apple commented:
This looks good, but I am not checking for exact correctness of keys, etc.
